content/advent-2017/mphf.md

```diff
@@ -0,0 +1,269 @@
+valid command before dispatching to the appropriate processing loop.
+
+```
+INFO
+CONN
+PUB 
+SUB 
+UNSU
+PING
+PONG
++OK 
+-ERR
+AUTH
+PUSH
+ADD 
+DECR
+SET 
+GET 
+QUIT
+```
+
+I’ve left the spaces after the three letter commands so that each key fits
+into a uint32.
+
+Here’s our first hash function. All it does it take the string as a 4-byte
+integer and takes the bottom 4 bits.
+
+```
+func hash1(x string) uint32 {
+    return binary.LittleEndian.Uint32([]byte(x)) & 15
+}
+```
+
+And then check for duplicates:
+
+```
+dups := make(map[uint32]bool)
+for _, k := range keys {
+    h := hash1(k)
+    if dups[h] {
+        log.Println("duplicate found")
+    }
+	dups[h] = true
+}
+```
+
+This is very fast, but when we test half of the keys collide. Changing the
+function to look at the upper 4 bits doesn't work either. The duplicate
+initial letters (PUSH, PUB) and trailing letters (PONG, PING) means we need
+to more shuffling.
+
+Here’s a second attempt:
+
+```
+const multiplier = 31
+func hash2(x string) int {
+   var u32 = (u32s(x) * multiplier) >> 28
+   return u32 & 15
+}
+```
+
+Here we’ve made two changes. First, we’re multiplying by 31, a nice random
+number that shows up in hash functions. Second, we’re going to extract the
+high bits of the result. High bits of multiplications tend to have a bit more
+entropy than the low bits, another common hash function trick.
+
+This is an improvement, although there are still 6 collisions, down from 8.
+But now we have a framework we can use. Can we find a value for `multiplier`
+that eliminates all collisions? This is easy enough to brute force.
+
+```
+2017/11/29 22:18:25 len(keys)= 16
+2017/11/29 22:18:25 searching for multiplier to fit 16 keys in 16 slots
+m = 715138
+INFO 3
+CONN 6
+PUB  15
+SUB  7
+UNSU 13
+PING 8
+PONG 0
++OK  1
+-ERR 4
+AUTH 2
+PUSH 11
+ADD  14
+DECR 5
+SET  10
+GET  9
+QUIT 12
+```
+
+And indeed, when we set the hash function to use `715138` we
+see that it does indeed map the 16 keys perfectly into 0 to 15.
+
+Lets benchmark this against a regular Go map.
+
+```
+var sink uint32
+
+func BenchmarkHash(b *testing.B) {
+	for i := 0; i < b.N; i++ {
+		for _, k := range keys {
+			sink += hash3(k)
+		}
+	}
+}
+
+func BenchmarkMap(b *testing.B) {
+	m := make(map[string]uint32)
+	for i, k := range keys {
+		m[k] = uint32(i)
+	}
+
+	b.ResetTimer()
+
+	for i := 0; i < b.N; i++ {
+		for _, k := range keys {
+			sink += m[k]
+		}
+	}
+}
+```
+
+And the output:
+
+```
+BenchmarkHash-8   	10000000	       126 ns/op
+BenchmarkMap-8    	10000000	       212 ns/op
+```
+
+About 40% faster to look up all the keys.
+
+But how can we generalize this? And does this always work? And is it always
+fast?
+
+Well, the first thing we notice is that as the set becomes larger, it becomes
+much more difficult to find a value for `multiplier` that works, and there
+might not even exist one. Even with 32 elements, trying to find the right
```

Comment:
"there might not even exist one" is a bit cumbersome, "one might not even exist"?